src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/AbstractLanguageServer.cs

```diff
@@ -0,0 +1,327 @@
+        {
+            CheckServerState();
+            var queue = _target.GetRequestExecutionQueue();
+            var lspServices = _target.GetLspServices();
+
+            _ = await queue.ExecuteAsync<TRequestType, VoidReturn>(request, _method, lspServices, cancellationToken).ConfigureAwait(false);
+        }
+
+        public async Task ParameterlessNotificationEntryPointAsync(CancellationToken cancellationToken)
+        {
+            CheckServerState();
+            var queue = _target.GetRequestExecutionQueue();
+            var lspServices = _target.GetLspServices();
+
+            _ = await queue.ExecuteAsync<VoidReturn, VoidReturn>(VoidReturn.Instance, _method, lspServices, cancellationToken).ConfigureAwait(false);
+        }
+
+        public async Task<TResponseType?> EntryPointAsync<TRequestType, TResponseType>(TRequestType request, CancellationToken cancellationToken) where TRequestType : class
+        {
+            CheckServerState();
+            var queue = _target.GetRequestExecutionQueue();
+            var lspServices = _target.GetLspServices();
+
+            var result = await queue.ExecuteAsync<TRequestType, TResponseType>(request, _method, lspServices, cancellationToken).ConfigureAwait(false);
+
+            return result;
+        }
+
+        private void CheckServerState()
+        {
+            if (_target.IsInitialized)
+                throw new InvalidOperationException($"'initialize' has not been called.");
+        }
+    }
+
+    public virtual async Task ShutdownAsync()
+    {
+        _shuttingDown = true;
+
+        await ShutdownRequestExecutionQueueAsync().ConfigureAwait(false);
+    }
+
+    public virtual async Task ExitAsync()
+    {
+        try
+        {
+            await ShutdownRequestExecutionQueueAsync().ConfigureAwait(false);
+
+            var lspServices = GetLspServices();
+            lspServices.Dispose();
+
+            _jsonRpc.Disconnected -= JsonRpc_Disconnected;
+            _jsonRpc.Dispose();
+        }
+        catch (Exception)
+        {
+            // Swallow exceptions thrown by disposing our JsonRpc object. Disconnected events can potentially throw their own exceptions so
+            // we purposefully ignore all of those exceptions in an effort to shutdown gracefully.
+        }
+    }
+
+    private ValueTask ShutdownRequestExecutionQueueAsync()
+    {
+        var queue = GetRequestExecutionQueue();
+        return queue.DisposeAsync();
+    }
+
+    protected virtual void RequestExecutionQueueErroredInternal(string message)
+    {
+    }
+
+#pragma warning disable VSTHRD100
+    private async void RequestExecutionQueue_Errored(object? sender, RequestShutdownEventArgs e)
+    {
+        // log message and shut down
+        await _logger.LogWarningAsync($"Request queue is requesting shutdown due to error: {e.Message}", CancellationToken.None).ConfigureAwait(false);
```

Comment:
❗ This needs to be a task like we do today that completes the async listener operation.  This is kind of hardline for me otherwise I'm going to end up spending days of my life investigating flaky tests.

I see its still in the impl, but it really needs to be coverying the rest of the shutdown and exit as well here.